### DISCUSSION TITLE
[Support] 循环依赖需要 forwardRef 支持

### DISCUSSION THREAD

**Discussion by fanioc** [midwayjs/midway#4051] [Q&A]
### Describe the problem（描述问题）

在 sevice1 和 service2 互相引用的情况下，会发生报错：
![image](https://github.com/user-attachments/assets/6cec2b9f-1288-449b-80de-2a57b1eaf647)
两个互相引用的 Service 写法如下：
```ts
// user.service.ts
import { Provide, Inject, Singleton } from '@midwayjs/core';
import { IUserOptions } from '../interface';
import { UserService2 } from './user2.service';

@Singleton()
@Provide()
export class UserService {
  @Inject()
  u2: UserService2;

  async getUser(options: IUserOptions) {
    return {
      uid: options.uid,
      username: 'mockedName',
      phone: '12345678901',
      email: 'xxx.xxx@xxx.com',
    };
  }
}
```
```ts
// user2.service.ts
import { Inject, Provide, Singleton } from '@midwayjs/core';
import { UserService } from './user.service';

@Singleton()
@Provide()
export class UserService2 {
  @Inject()
  u1: UserService;
}
```

这里是编译后的代码，set metadata 执行时 `user_service_1.UserService` 的值是 undefined：

<img width="675" alt="image" src="https://github.com/user-attachments/assets/2b50238f-2b38-4d7d-8e37-907499c246cc">

因为`emitDecoratorMetadata`编译后的 set metadata 是立即执行的，在两个模块发生循环引用时，立即执行的代码拿到的值就是undefied，所以 set metadata 也是 undefined，导致无法正确读取到依赖注入的对象。

所以需要一个像Nestjs 中的 **`@Inject(forwardRef(()=>UserService2))`** 方式来解决循环引用时变量初始化为 undefiend 情况，只解决框架的循环依赖逻辑还不够，有一个相关的 issue https://github.com/midwayjs/midway/issues/3197

### Midway Versions（Midway 版本）

>> Start to check your midway component version...

✓ @midwayjs/faas-typings(not installed)
✓ @midwayjs/fc-starter(not installed)
✓ @midwayjs/serverless-http-parser(not installed)
✓ @midwayjs/async-hooks-context-manager(not installed)
✓ @midwayjs/axios(not installed)
✓ @midwayjs/bootstrap(3.16.6)
✓ @midwayjs/bull(not installed)
✓ @midwayjs/bull-board(not installed)
✓ @midwayjs/cache-manager(not installed)
✓ @midwayjs/captcha(not installed)
✓ @midwayjs/casbin(not installed)
✓ @midwayjs/casbin-redis-adapter(not installed)
✓ @midwayjs/casbin-typeorm-adapter(not installed)
✓ @midwayjs/code-dye(not installed)
✓ @midwayjs/consul(not installed)
✓ @midwayjs/core(3.16.2)
✓ @midwayjs/cos(not installed)
✓ @midwayjs/cron(not installed)
✓ @midwayjs/cross-domain(not installed)
✓ @midwayjs/decorator(not installed)
✓ @midwayjs/etcd(not installed)
✓ @midwayjs/express-session(not installed)
✓ @midwayjs/faas(not installed)
✓ @midwayjs/grpc(not installed)
✓ @midwayjs/http-proxy(not installed)
✓ @midwayjs/i18n(not installed)
✓ @midwayjs/info(3.16.8)
✓ @midwayjs/jwt(not installed)
✓ @midwayjs/kafka(not installed)
✓ @midwayjs/leoric(not installed)
✓ @midwayjs/mikro(not installed)
✓ @midwayjs/mock(3.16.5)
✓ @midwayjs/mongoose(not installed)
✓ @midwayjs/mqtt(not installed)
✓ @midwayjs/oss(not installed)
✓ @midwayjs/otel(not installed)
✓ @midwayjs/passport(not installed)
✓ @midwayjs/process-agent(not installed)
✓ @midwayjs/prometheus(not installed)
✓ @midwayjs/prometheus-socket-io(not installed)
✓ @midwayjs/rabbitmq(not installed)
✓ @midwayjs/redis(not installed)
✓ @midwayjs/security(not installed)
✓ @midwayjs/sequelize(not installed)
✓ @midwayjs/session(not installed)
✓ @midwayjs/socketio(not installed)
✓ @midwayjs/static-file(not installed)
✓ @midwayjs/swagger(not installed)
✓ @midwayjs/tablestore(not installed)
✓ @midwayjs/tags(not installed)
✓ @midwayjs/tenant(not installed)
✓ @midwayjs/typegoose(not installed)
✓ @midwayjs/typeorm(not installed)
✓ @midwayjs/upload(not installed)
✓ @midwayjs/validate(3.16.8)
✓ @midwayjs/version(not installed)
✓ @midwayjs/view(not installed)
✓ @midwayjs/view-ejs(not installed)
✓ @midwayjs/view-nunjucks(not installed)
✓ @midwayjs/web(not installed)
✓ @midwayjs/express(not installed)
✓ @midwayjs/koa(3.16.8)
✓ @midwayjs/ws(not installed)
**********************************************************************
>> Check complete, all versions are healthy.
**********************************************************************

**Comment by czy88840616** [comment]
循环的地方用动态 API 获取就好。

**Reply by fanioc** [reply]
动态api获取，个人感觉没有Inject(()=>UserService)直观，也不是常规手段，大型项目比较容易出现循环引用的问题，期望针对undefined获取注入实例的情况，在框架提示一下可能存在循环引用，不然排查问题比较痛苦

**Reply by czy88840616** [reply]
框架这个地方其实已经做了，如果碰到循环引用，现在应该会变成 proxy 对象，你这个 case 其实我测下来没有问题，不过你说的的确存在，目前在 v3 不会在核心这个地方做改动，v4 会处理。

**Reply by flyingcrp** [reply]
关注一下，这个问题确实很实在实用
